### PR TITLE
feat(agw): add ho algorithm parameter

### PR DESCRIPTION
**Pull request by sxkmg**
Signed-off-by: root <369888042@qq.com>

 feat(agw): add ho algorithm parameter
   

## Summary

add None 243 Ho  algorithm parameter

## Test Plan

 ##1.Start the magma service
2.Configure the base station none-243 to connect to the gateway
3.Test and issue parameters through the magma swagger API(https://magma-test.localhost:9443/swagger/v1/ui/lte#/EnodeBs/get_lte__network_id__enodebs__enodeb_serial_)
1631007352343_F143C943-8C85-4825-BC5E-BAD96161522E
4.View the configuration file and verify whether the parameters are distributed to the gateway
（ sudo tail -1000f /var/opt/magma/configs/gateway.mconfig ）
lQLPDhqt8XIw6jjNAlzNBGiwiJoxtZlFWOSixRPGnvC3pQ_1128_604

5.Check whether the parameters are successfully distributed to the base station
sudo tail -1000f /var/log/enodebd.log
lALPD4PvOODaWJHNAZXNBms_1643_405

image

### CODE REVIEW

File: lte/protos/mconfig/mconfigs.proto [hunk 1]
```diff
@@ -71,6 +71,33 @@ message EnodebD {
...
+        int32 qrxlevmin_sib1 = 105;
+        uint32 qrxlevminoffset = 106;
+        uint32 s_intrasearch = 107;
+        uint32 s_nonintrasearch = 108;
+        int32 qrxlevmin_sib3 = 109;
+        uint32 reselection_priority = 110;
+        uint32 threshservinglow = 111;
+        bool x2_enable_disable = 112;
+        string ciphering_algorithm = 113;
+        string integrity_algorithm = 114;
```
Comment (uri200) [inline lte/protos/mconfig/mconfigs.proto]: please, create a message to encapsulate this config

```
message HoAlgorithmConfig {
		int32 a1_threshold_rsrp = 1;
        int32 lte_a1_threshold_rsrq = 2;
        int32 hysteresis = 3;
        string time_to_trigger = 4;
		...
}
```

File: lte/cloud/go/services/lte/servicers/builder_servicer.go [hunk 2]
```diff
@@ -397,6 +397,37 @@ func getEnodebConfigsBySerial(nwConfig *lte_models.NetworkCellularConfigs, gwCon
...
+				enbMconfig.Qrxlevminoffset = uint32(hoAlgorithmConfig.Qrxlevminoffset)
+				enbMconfig.SIntrasearch = uint32(hoAlgorithmConfig.SIntrasearch)
+				enbMconfig.SNonintrasearch = uint32(hoAlgorithmConfig.SNonintrasearch)
+				enbMconfig.QrxlevminSib3 = int32(hoAlgorithmConfig.QrxlevminSib3)
+				enbMconfig.ReselectionPriority = uint32(hoAlgorithmConfig.ReselectionPriority)
+				enbMconfig.Threshservinglow = uint32(hoAlgorithmConfig.Threshservinglow)
+				enbMconfig.X2EnableDisable = swag.BoolValue(hoAlgorithmConfig.X2EnableDisable)
+				enbMconfig.CipheringAlgorithm = string(hoAlgorithmConfig.CipheringAlgorithm)
+				enbMconfig.IntegrityAlgorithm = string(hoAlgorithmConfig.IntegrityAlgorithm)
+
```
Comment (uri200) [inline lte/cloud/go/services/lte/servicers/builder_servicer.go]: please, put this code in a function like
```
func getEnodeBHoAlgorithmConfign() 
```

**Review by uri200 — CHANGES_REQUESTED** [review]

**Comment by uri200** [comment]
it looks like you will need to run the test first

For orc8r
```
cd ~/magma/orc8r/cloud/docker
./build -g
./build -c
```

For agw
```
cd ~/magma/lte/gateway
vagrant up magma
vagrant ssh magma
# inside vagrant
cd magma/lte/gateway
make 
```

File: lte/cloud/go/services/lte/servicers/builder_servicer.go (hunk 2 above)
Comment (sxkmg) [reply lte/cloud/go/services/lte/servicers/builder_servicer.go]: Good idea, optimized, thank you

File: lte/protos/mconfig/mconfigs.proto (hunk 1 above)
Comment (sxkmg) [reply lte/protos/mconfig/mconfigs.proto:100]: image

Comment (sxkmg) [reply lte/protos/mconfig/mconfigs.proto]: image

File: lte/cloud/go/services/lte/servicers/builder_servicer.go (hunk 2 above)
Comment (sxkmg) [reply lte/cloud/go/services/lte/servicers/builder_servicer.go]: lQLPDhqt8uDCd6fNAmPNBDmwshqxN_nVTihM59F9jUAHEQ_1081_611

Comment (sxkmg) [reply lte/cloud/go/services/lte/servicers/builder_servicer.go]: Please review again, thank you！

**Comment by sxkmg** [comment]
The suggested modification has been completed, please review again, thank you!
lQLPDhqt8uDCd6fNAmPNBDmwshqxN_nVTihM59F9jUAHEQ_1081_611
lQLPDhqt8uFbDgHNAULNA1OwoOFl6ZB9gzcvh-QH4Z8luA_851_322

**Comment by uri200** [comment]
Thanks for the changes. Golang part seems fine to me now.

**Review by uri200 — APPROVED** [review]

**Comment by uri200** [comment]
@andreilee @pshelar please take a look at the NMS and python parts

**Comment by uri200** [comment]
Hey @sxkmg  can you please fix the linter errors and rebase?

File: orc8r/cloud/docker/.env [hunk 3]
```diff
@@ -10,3 +10,4 @@
 # limitations under the License.
 
 COMPOSE_PROJECT_NAME=orc8r
+COMPOSE_FILE=docker-compose.yml:docker-compose.metrics.yml:docker-compose.override.yml  # DO NOT COMMIT THIS CHANGE
```
Comment (talkhasib) [inline orc8r/cloud/docker/.env]: Does this need to be cleaned?

Comment (sxkmg) [reply orc8r/cloud/docker/.env]: ok，thank you!

Comment (sxkmg) [reply orc8r/cloud/docker/.env]: The file has changed，thanks

File: go.mod [hunk 4]
```diff
@@ -0,0 +1,3 @@
+module magma
+
+go 1.16
```
Comment (uri200) [inline go.mod:3]: Why is this file needed here?

Note that we can;t switch to go 1.16

**Review by uri200 — CHANGES_REQUESTED** [review]

File: orc8r/tools/fab/dev_utils.py [hunk 5]
```diff
@@ -1,9 +1,7 @@
 """
 Copyright 2020 The Magma Authors.
-
```
Comment (uri200) [inline orc8r/tools/fab/dev_utils.py:3]: There are not significative changes here. Can you un stage this file please

File: lte/cloud/go/services/lte/obsidian/models/swagger.v1.yml [hunk 6]
```diff
@@ -2305,7 +2305,118 @@ definitions:
           - 'NuRAN Cavium OC-LTE'
           - 'FreedomFi One'
         x-nullable: false
+      ho_algorithm_config:
```
Comment (andreilee) [inline lte/cloud/go/services/lte/obsidian/models/swagger.v1.yml:2463]: This should be explicitly marked as optional. Please add some details on specifications that other developers or API users could look up to understand how to use these fields correctly.

File: lte/cloud/go/services/lte/obsidian/models/swagger.v1.yml [hunk 7]
```diff
@@ -2305,7 +2305,118 @@ definitions:
...
+        example: "ms480"
+      a2_threshold_rsrp:
+        type: integer
+        format: int32
+        example: 5
+      lte_a2_threshold_rsrq:
+        type: integer
+        format: int32
+        example: 6
+      lte_a2_threshold_rsrp_irat_volte:
```
Comment (andreilee) [inline lte/cloud/go/services/lte/obsidian/models/swagger.v1.yml:2493]: All of these fields should have some more detailed description, value limits (min, max, or enum, etc.) as well as references to specifications to make the definitions easy to understand, and unambiguous. Many of these fields can also be further sub-grouped.

File: lte/gateway/python/magma/enodebd/device_config/configuration_init.py [hunk 8]
```diff
@@ -86,6 +99,34 @@ def build_desired_config(
     # Attempt to load device configuration from YANG before service mconfig
     enb_config = _get_enb_yang_config(device_config) or \
                  _get_enb_config(mconfig, device_config)
+    # Ho Algorithm
+    _set_algorithm_a1_threshold_rsrp(cfg_desired, enb_config.a1_threshold_rsrp)
+    # _set_algorithm_lte_a1_threshold_rsrq(cfg_desired, enb_config.lte_a1_threshold_rsrq)
+    # _set_algorithm_hysteresis(cfg_desired, enb_config.hysteresis)
+    # _set_algorithm_time_to_trigger(cfg_desired, enb_config.time_to_trigger)
```
Comment (andreilee) [inline lte/gateway/python/magma/enodebd/device_config/configuration_init.py:106]: Why are these lines commented? Should they just be removed? Otherwise add some other comments to explain.

File: lte/protos/mconfig/mconfigs.proto [hunk 9]
```diff
@@ -71,6 +71,37 @@ message EnodebD {
         int32 tac = 8;
         int32 cell_id = 9;
         string ip_address = 10;
+        HoAlgorithmConfig ho_algorithm_config = 16;
+        message HoAlgorithmConfig {
+            // ho algorithm config
```
Comment (andreilee) [inline lte/protos/mconfig/mconfigs.proto:77]: Add some more detailed description here - is there a link to a specification? This comment is just repeating what the name `HoAlgorithmConfig` already tells us. Another developer coming in to use the definitions you've established here can interpret these fields in slightly different ways - please add as much clarity here as possible.

File: nms/packages/magmalte/app/views/equipment/EnodebDetailConfigEdit.js [hunk 10]
```diff
@@ -472,6 +579,329 @@ export function RanEdit(props: Props) {
...
+                  value={optConfig.b2_threshold1_rsrp}
+                  onChange={({target}) =>
+                    handleOptChange('b2_threshold1_rsrp', target.value)
+                  }
+                />
+              </AltFormField>
+              <AltFormField label={'b2 threshold2 rsrp'}>
+                <OutlinedInput
+                  data-testid="b2_threshold2_rsrp"
+                  placeholder="Enter B2THRESHOLD2RSRP"
```
Comment (andreilee) [inline nms/packages/magmalte/app/views/equipment/EnodebDetailConfigEdit.js:779]: Is there a particular reason for the all-caps placeholders? I don't imagine that this is particularly legible.

**Review by andreilee — CHANGES_REQUESTED** [review]
Thanks for submitting this pull request and including a detailed test plan! While this adds a lot more configurability for eNodeBs through Magma, I think there are some code legibility improvements that can be made, and also some UX improvements.

File: lte/gateway/python/magma/enodebd/devices/baicells_rts.py [hunk 11]
```diff
@@ -221,6 +221,124 @@ class BaicellsRTSTrDataModel(DataModel):
...
+        ),
+        ParameterName.HYSTERESIS: TrParam(
+            FAPSERVICE_PATH + 'CellConfig.LTE.RAN.Mobility.ConnMode.EUTRA.A1ThresholdRSRP', True, TrParameterType.INT,
+            False,
+        ),
+        ParameterName.TIME_TO_TRIGGER: TrParam(
+            FAPSERVICE_PATH + 'CellConfig.LTE.RAN.Mobility.ConnMode.EUTRA.A1ThresholdRSRP', True,
+            TrParameterType.STRING,
+            False,
+        ),
```
Comment (talkhasib) [inline lte/gateway/python/magma/enodebd/devices/baicells_rts.py:241]: How come all of these map to the same param "A1ThresholdRSRP"?

File: lte/gateway/python/magma/enodebd/devices/baicells_rts.py [hunk 12]
```diff
@@ -221,6 +221,124 @@ class BaicellsRTSTrDataModel(DataModel):
...
+        ParameterName.LTE_A2_THRESHOLD_RSRP_IRAT_VOLTE: TrParam(
+            FAPSERVICE_PATH + 'CellConfig.LTE.RAN.Mobility.ConnMode.EUTRA.A2ThresholdRSRP', True, TrParameterType.INT,
+            False,
+        ),
+        ParameterName.LTE_A2_THRESHOLD_RSRQ_IRAT_VOLTE: TrParam(
+            FAPSERVICE_PATH + 'CellConfig.LTE.RAN.Mobility.ConnMode.EUTRA.A2ThresholdRSRP', True, TrParameterType.INT,
+            False,
+        ),
```
Comment (talkhasib) [inline lte/gateway/python/magma/enodebd/devices/baicells_rts.py:257]: Same here, 4 params map to the same config?

File: lte/gateway/python/magma/enodebd/device_config/configuration_init.py [hunk 13]
```diff
@@ -86,6 +99,34 @@ def build_desired_config(
...
+    _set_algorithm_b2_geran_irat_threshold(cfg_desired, enb_config.b2_geran_irat_threshold)
+    _set_algorithm_qrxlevmin_sib1(cfg_desired, enb_config.qrxlevmin_sib1)
+    _set_algorithm_qrxlevminoffset(cfg_desired, enb_config.qrxlevminoffset)
+    _set_algorithm_s_intrasearch(cfg_desired, enb_config.s_intrasearch)
+    _set_algorithm_s_nonintrasearch(cfg_desired, enb_config.s_nonintrasearch)
+    _set_algorithm_qrxlevmin_sib3(cfg_desired, enb_config.qrxlevmin_sib3)
+    _set_algorithm_reselection_priority(cfg_desired, enb_config.reselection_priority)
+    _set_algorithm_threshservinglow(cfg_desired, enb_config.threshservinglow)
+    _set_algorithm_ciphering_algorithm(cfg_desired, enb_config.ciphering_algorithm)
+    _set_algorithm_integrity_algorithm(cfg_desired, enb_config.integrity_algorithm)
```
Comment (talkhasib) [inline lte/gateway/python/magma/enodebd/device_config/configuration_init.py:129]: Combine all of these into a single "_set_handover_params()" function

**Review by talkhasib — CHANGES_REQUESTED** [review]

**Comment by talkhasib** [comment]
Few high level things need to happen for this PR to move forward
A) The PR needs to be split into at least 4 parts (in the order below)
   1- Add HO parameters to enodeb data model in enodenbd
   2- Consume static agw config and map to data model
   3- Add config to swagger API
   4- Add NMS UI

B) Vendor agnostic config through NMS and API
We should differentiate between HW/Vendor specific configuration and global, vendor-agnostic, config.
The HO config added in this PR is an exact replica of the HW config. Including the naming of config params.
We should look at HO config across multiple Baicells models as well as across other vendors to abstract the high level config across all vendors. We also should structure the config and use names representative of the functionality

File: lte/gateway/python/magma/enodebd/data_models/data_model_parameters.py [hunk 14]
```diff
@@ -85,6 +85,35 @@ class ParameterName():
...
+    QRXLEVMIN_SIB1 = 'qrxlevmin sib1'
+    QRXLEVMINOFFSET = 'qrxlevminoffset'
+    S_INTRASEARCH = 's intrasearch'
+    S_NONINTRASEARCH = 's nonintrasearch'
+    QRXLEVMIN_SIB3 = 'qrxlevmin sib3'
+    RESELECTION_PRIORITY = 'reselection priority'
+    THRESHSERVINGLOW = 'threshservinglow'
+    CIPHERING_ALGORITHM = 'ciphering algorithm'
+    INTEGRITY_ALGORITHM = 'integrity algorithm'
+
```
Comment (talkhasib) [inline lte/gateway/python/magma/enodebd/data_models/data_model_parameters.py]: prefix all of these with HO_
Also, many of these parameters are not self explanatory. Either add detailed commenting or improve param naming
Also, instead of using Sib1/Sib3, use cell selection/reselection terminology
Same for units used by each, db vs linear ...

File: lte/gateway/python/magma/enodebd/device_config/configuration_init.py [hunk 15]
```diff
@@ -48,6 +48,19 @@
...
+        'lte_a2_threshold_rsrp_irat_volte', 'lte_a2_threshold_rsrq_irat_volte',
+        'a3_offset', 'a3_offset_anr',
+        'a4_threshold_rsrp', 'lte_intra_a5_threshold_1_rsrp',
+        'lte_intra_a5_threshold_2_rsrp', 'lte_inter_anr_a5_threshold_1_rsrp',
+        'lte_inter_anr_a5_threshold_2_rsrp', 'b2_threshold1_rsrp',
+        'b2_threshold2_rsrp', 'b2_geran_irat_threshold',
+        'qrxlevmin_sib1', 'qrxlevminoffset',
+        's_intrasearch', 's_nonintrasearch',
+        'qrxlevmin_sib3', 'reselection_priority',
+        'threshservinglow', 'ciphering_algorithm', 'integrity_algorithm',
```
Comment (talkhasib) [inline lte/gateway/python/magma/enodebd/device_config/configuration_init.py:63]: If we are going to be adding this large number of params. We might want to reconsider using a namedtuple and use a dictionary with some hierarchy

File: magma.iml [hunk 16]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (talkhasib) [inline magma.iml:1]: What is this file? Why is it added?

**Review by talkhasib — CHANGES_REQUESTED** [review]

File: go.mod (hunk 4 above)
Comment (sxkmg) [reply go.mod:3]: The file automatically generated by ide was submitted by mistake and has been deleted,thank you .

File: lte/gateway/python/magma/enodebd/device_config/configuration_init.py (hunk 8 above)
Comment (sxkmg) [reply lte/gateway/python/magma/enodebd/device_config/configuration_init.py:106]: The parameters in this part do not provide a path, so they are commented out, like baicells_ rts. The duplicate mapping in py file is the same, and has been deleted at present。

File: lte/gateway/python/magma/enodebd/devices/baicells_rts.py (hunk 12 above)
Comment (sxkmg) [reply lte/gateway/python/magma/enodebd/devices/baicells_rts.py:257]: The parameters in this part do not provide a path. In order to keep the code and facilitate the use after providing a path, the same path is mapped temporarily and has been deleted at present.

File: lte/gateway/python/magma/enodebd/device_config/configuration_init.py (hunk 13 above)
Comment (sxkmg) [reply lte/gateway/python/magma/enodebd/device_config/configuration_init.py:129]: image
Functions have been integrated, thank you！

File: lte/gateway/python/magma/enodebd/device_config/configuration_init.py (hunk 15 above)
Comment (sxkmg) [reply lte/gateway/python/magma/enodebd/device_config/configuration_init.py:63]: image
Parameters have been integrated, thank you!

File: lte/gateway/python/magma/enodebd/data_models/data_model_parameters.py (hunk 14 above)
Comment (sxkmg) [reply lte/gateway/python/magma/enodebd/data_models/data_model_parameters.py]: image
image
Redefined and modified, thank you!

File: magma.iml (hunk 16 above)
Comment (sxkmg) [reply magma.iml:1]: Deleted, thank you!

File: lte/cloud/go/services/lte/obsidian/models/swagger.v1.yml (hunk 7 above)
Comment (sxkmg) [reply lte/cloud/go/services/lte/obsidian/models/swagger.v1.yml:2493]: 1640509171458_A45D7BBA-F021-4257-A860-7EA46E2D2A78
Modified, thank you!

File: lte/cloud/go/services/lte/obsidian/models/swagger.v1.yml (hunk 6 above)
Comment (sxkmg) [reply lte/cloud/go/services/lte/obsidian/models/swagger.v1.yml:2463]: 1640509029661_773D068C-61E9-461a-A918-38F399500364
Modified, thank you!

File: nms/packages/magmalte/app/views/equipment/EnodebDetailConfigEdit.js (hunk 10 above)
Comment (sxkmg) [reply nms/packages/magmalte/app/views/equipment/EnodebDetailConfigEdit.js:779]: image
Modified, thank you!

File: lte/gateway/python/magma/enodebd/devices/baicells_rts.py (hunk 11 above)
Comment (sxkmg) [reply lte/gateway/python/magma/enodebd/devices/baicells_rts.py:241]: The parameters in this part do not provide a path. In order to keep the code and facilitate the use after providing a path, the same path is mapped temporarily and has been deleted at present.

**Comment by hcgatewood** [comment]
Closing for now due to inactivity. Please feel free to reopen and message/DM if you'd like support in getting this merged 🙂

**Comment by sxkmg** [comment]
I look forward to reopening the current PR, thanks!
